File: commonplace/parse.py
import dataclasses
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Iterable, Tuple

from commonplace.line_iterator import LineIterator, StringLineIterator
from commonplace.models import (Category, Comment, DocPosition, Line, Moment, MomentDateTime, ParseConfig, Recurrence,
                                RecurrenceType, RecurringMoment, SingleMoment, Todos, WorkState)
from commonplace.util import epoch_week, with_end_of_day, with_weekday


@dataclass
class ParseState:
    config: ParseConfig
    line_iter: LineIterator
    todos: Todos


def parse_moments_file(path: str, config: ParseConfig) -> Todos:
    with open(path, "rU", encoding="utf8") as file:
        return _parse_moments(LineIterator(file), config)


def parse_moments_string(content: str, config: ParseConfig) -> Todos:
    return _parse_moments(StringLineIterator(content), config)


def parse_moments(lines: Iterable[str], config: ParseConfig) -> Todos:
    return _parse_moments(LineIterator(lines), config)


def _parse_moments(line_iter: LineIterator, config: ParseConfig) -> Todos:
    state = ParseState(config=config, line_iter=line_iter, todos=Todos())
    for line in _each_line(state.line_iter):
        _parse_line(line, state)

    return state.todos


def _parse_line(line: Line, state: ParseState):
    if _is_blank(line.content):
        return

    if _is_category_delimiter(line, state.config):
        _parse_category_block(state)
    elif line.content.startswith(state.config.left_state_bracket):
        _parse_top_moment_block(line, state)


def _parse_category_block(state: ParseState):
    cat_line = next(state.line_iter, None)
    if cat_line is None:
        # Expected a category name after category delimiter
        return

    line_after_cat = next(state.line_iter, None)
    if not _is_category_delimiter(line_after_cat, state.config):
        # Expected closing delimiter line
        return

    state.todos.categories.append(_parse_category(cat_line, state.config))


def _parse_category(line: Line, config: ParseConfig) -> Category:
    content = line.content
    color, content = _parse_category_color(content)
    priority, content = _parse_priority(content, config)
    return Category(
        name=content.lstrip(),
        color=color,
        priority=priority,
        doc_pos=DocPosition(
            line_num=line.line_num,
            offset=line.offset,
            length=len(line.content),
        ),
    )


def _parse_category_color(content: str) -> Tuple[str, str]:
    start = content.find("[")

    if start < 0 or not content.endswith("]"):
        return "", content

    return content[start + 1:-1], content[:start].strip()


def _parse_priority(content: str, config: ParseConfig) -> Tuple[int, str]:
    prio = 0
    for prio, char in enumerate(reversed(content)):
        if char != config.priority_mark:
            break

    return prio, content[:len(content) - prio].strip()


def _is_category_delimiter(line: Line, config: ParseConfig):
    return line is not None and line.content.startswith(config.category_delim)


def _parse_top_moment_block(line: Line, state: ParseState):
    mom = _parse_moment_block(line, line.content.strip(), state)
    if not mom:
        return

    state.todos.moments.append(mom)


def _parse_moment_block(line: Line, line_content: str, state: ParseState, indent=0):
    mom = _parse_moment_line(line, line_content, state.config)
    if not mom:
        return None

    mom.category = state.todos.categories[-1] if state.todos.categories else None
    _parse_comments_and_sub_moments(mom, state, indent)
    return mom


def _parse_moment_line(line: Line, line_content: str, config: ParseConfig) -> Moment | None:
    mom, line_content = _parse_recurring_moment(line, line_content, config)
    if not mom:
        mom, line_content = _parse_single_moment(line, line_content, config)

    state, line_content = _parse_state_mark(line_content, config)
    if state is None:
        return None
    mom.work_state = state

    mom.priority, line_content = _parse_priority(line_content, config)

    mom.name = line_content

    return mom


def _parse_recurring_moment(line: Line, line_content: str, config: ParseConfig) -> Tuple[RecurringMoment, str]:
    if not line_content.endswith(config.right_date_bracket):
        return None, line_content

    recur, time_of_day, new_line_content = _parse_recurrence(line, line_content, config)
    if not recur:
        return None, line_content

    mom = RecurringMoment(recurrence=recur,
                          time_of_day=time_of_day,
                          doc_pos=DocPosition(
                              line_num=line.line_num,
                              offset=line.offset,
                              length=len(line.content),
                          ))
    return mom, new_line_content


def _parse_recurrence(line: Line, line_content: str, config: ParseConfig) -> Tuple[Recurrence, MomentDateTime, str]:
    lbracket_pos = line_content.rfind(config.left_date_bracket)
    if lbracket_pos < 0:
        return None, None, line_content

    untrimmed_pos = line.content.rfind(config.left_date_bracket) + 1
    recur_str = line_content[lbracket_pos + 1:-1]
    time_of_day, recur_str = _parse_time_suffix(recur_str, config)
    if time_of_day:
        time_of_day.doc_pos.offset += line.offset + untrimmed_pos

    recur = None
    for parse in [try_parse_daily, try_parse_weekly, try_parse_n_weekly, try_parse_monthly, try_parse_yearly]:
        recur = parse(recur_str, config)
        if recur:
            break

    if not recur:
        return None, None, line_content

    recur.ref_date.doc_pos = DocPosition(
        line_num=line.line_num,
        offset=line.offset + untrimmed_pos,
        length=len(recur_str),
    )

    return recur, time_of_day, line_content[:lbracket_pos].strip()


def try_parse_daily(content: str, config: ParseConfig) -> Recurrence | None:
    if not config.daily_pattern.search(content):
        return None

    return Recurrence(
        recurrence_type=RecurrenceType.DAILY,
        ref_date=MomentDateTime(dt=config.now()),
    )


def try_parse_weekly(content: str, config: ParseConfig) -> Recurrence | None:
    match = config.weekly_pattern.search(content)
    if not match:
        return None

    weekday = _parse_weekday(match.group(1), config)
    if weekday < 0:
        return None

    ref_date = with_weekday(config.now(), weekday)
    return Recurrence(
        recurrence_type=RecurrenceType.WEEKLY,
        ref_date=MomentDateTime(dt=_with_start_of_day(ref_date.astimezone(tz=None))),
    )


def _parse_weekday(content: str, config: ParseConfig) -> int:
    return config.week_days.index(content.lower())


def _with_start_of_day(date: datetime) -> datetime:
    return date.replace(hour=0, minute=0, second=0, microsecond=0)


def try_parse_n_weekly(content: str, config: ParseConfig) -> Recurrence | None:
    match = config.n_weekly_pattern.search(content)
    if not match:
        return None

    nth, nth_recur_type = _parse_nth(match.group(1), config)
    if not nth_recur_type:
        return None

    weekday = _parse_weekday(match.group(2), config)
    if weekday < 0:
        return None

    ref_date = with_weekday(config.now(), weekday)
    week_offset = epoch_week(ref_date) % nth
    ref_date += timedelta(days=-7 * week_offset)
    return Recurrence(
        recurrence_type=nth_recur_type,
        ref_date=MomentDateTime(dt=_with_start_of_day(ref_date.astimezone(tz=None))),
    )


def _parse_nth(content: str, config: ParseConfig) -> Tuple[int, RecurrenceType]:
    nth = config.nths.index(content.lower())
    if nth == 0:
        return 2, RecurrenceType.BI_WEEKLY
    if nth == 1:
        return 3, RecurrenceType.TRI_WEEKLY
    if nth == 2:
        return 4, RecurrenceType.QUADRI_WEEKLY

    return None, None


def try_parse_monthly(content: str, config: ParseConfig) -> Recurrence | None:
    match = config.monthly_pattern.search(content)
    if not match:
        return None

    try:
        day = int(match.group(1))
    except ValueError:
        return None

    now = config.now()
    ref_date = datetime(year=now.year, month=now.month, day=day, hour=0, minute=0, second=0, microsecond=0)
    return Recurrence(
        recurrence_type=RecurrenceType.MONTHLY,
        ref_date=MomentDateTime(dt=ref_date.astimezone(tz=None)),
    )


def try_parse_yearly(content: str, config: ParseConfig) -> Recurrence | None:
    match = config.yearly_pattern.search(content)
    if not match:
        return None

    try:
        day = int(match.group(1))
        month = int(match.group(2))
    except ValueError:
        return None

    now = config.now()
    ref_date = datetime(year=now.year, month=month, day=day, hour=0, minute=0, second=0, microsecond=0)
    return Recurrence(
        recurrence_type=RecurrenceType.YEARLY,
        ref_date=MomentDateTime(dt=ref_date.astimezone(tz=None)),
    )


def _parse_single_moment(line: Line, line_content: str, config: ParseConfig) -> Tuple[SingleMoment, str]:
    start = None
    end = None
    time_of_day = None
    if line_content.endswith(config.right_date_bracket):
        start, end, time_of_day, line_content = _parse_date_suffix(line, line_content, config)

    return SingleMoment(start=start,
                        end=end,
                        time_of_day=time_of_day,
                        doc_pos=DocPosition(
                            line_num=line.line_num,
                            offset=line.offset,
                            length=len(line.content),
                        )), line_content


def _parse_date_suffix(line: Line, line_content: str,
                       config: ParseConfig) -> Tuple[MomentDateTime, MomentDateTime, MomentDateTime, str]:
    left_pos = line_content.rfind(config.left_date_bracket)
    if left_pos < 0:
        return None, None, None, line_content

    untrimmed_pos = line.content.rfind(config.left_date_bracket) + 1
    suffix_str = line_content[left_pos + 1:-1]
    time_of_day, suffix_str = _parse_time_suffix(suffix_str, config)
    _finalize_doc_pos(time_of_day, line, untrimmed_pos)

    suffix_str = suffix_str.strip()
    start = None
    end = None
    # Try all dashes as separator in case the date format uses dashes (e.g. 2015-12-24 - 2016-02-03)
    dash_offset = 0
    dash_pos = suffix_str.find("-")
    while dash_pos >= 0 and start is None:
        start, end = _parse_date_range(suffix_str, dash_offset + dash_pos, config)
        if start is None:
            dash_offset += dash_pos + 1
            dash_pos = suffix_str.find("-", dash_offset)

    if not start:
        start = _parse_date_single(suffix_str, config)
        if start:
            end = dataclasses.replace(start)
            end.doc_pos = dataclasses.replace(start.doc_pos)

    if end:
        end.dt = with_end_of_day(end.dt)

    if start or end:
        _finalize_doc_pos(start, line, untrimmed_pos)
        _finalize_doc_pos(end, line, untrimmed_pos)
        return start, end, time_of_day, line_content[:left_pos].strip()

    return None, None, None, line_content


def _parse_date_range(line_content: str, dash_pos: int, config: ParseConfig) -> Tuple[MomentDateTime, MomentDateTime]:
    start: MomentDateTime | None = None
    end: MomentDateTime | None = None
    start_str = line_content[:dash_pos]
    end_str = line_content[dash_pos + 1:]

    if start_str:
        date = _parse_date(start_str, config)
        if not date:
            return None, None
        start = MomentDateTime(dt=date,
                               doc_pos=DocPosition(
                                   line_num=-1,
                                   offset=_count_start_whitespaces(start_str),
                                   length=_stripped_length(start_str),
                               ))

    if end_str:
        date = _parse_date(end_str, config)
        if not date:
            return None, None
        end = MomentDateTime(dt=date,
                             doc_pos=DocPosition(
                                 line_num=-1,
                                 offset=len(start_str) + 1 + _count_start_whitespaces(end_str),
                                 length=_stripped_length(end_str),
                             ))

    return start, end


def _parse_date_single(line_content: str, config: ParseConfig) -> MomentDateTime | None:
    date = _parse_date(line_content, config)

    if not date:
        return None

    return MomentDateTime(dt=date,
                          doc_pos=DocPosition(
                              line_num=-1,
                              offset=_count_start_whitespaces(line_content),
                              length=_stripped_length(line_content),
                          ))


def _parse_time_suffix(line_content: str, config: ParseConfig) -> Tuple[MomentDateTime, str]:
    trimmed = line_content.strip()
    space_pos = trimmed.rfind(" ")
    if space_pos < 0 or space_pos == len(trimmed) - 1:
        return None, line_content

    time_str = trimmed[space_pos + 1:]
    time_of_day = _parse_time(time_str, config)
    if not time_of_day:
        return None, line_content

    mom_time = MomentDateTime(dt=time_of_day, doc_pos=DocPosition(
        line_num=-1,
        offset=space_pos + 1,
        length=len(time_str),
    ))

    return mom_time, line_content[:space_pos]


def _finalize_doc_pos(date: MomentDateTime, line: Line, offset_delta):
    if date:
        date.doc_pos.line_num = line.line_num
        date.doc_pos.offset += line.offset + offset_delta


def _parse_state_mark(line_content: str, config: ParseConfig) -> Tuple[WorkState, str]:  # pylint: disable=too-many-return-statements
    match_res = config.state_mark_pattern.match(line_content)
    if not match_res:
        return None, line_content

    leftover = line_content[match_res.end() + 1:].strip()
    return config.state_marks.get(match_res.group(1), WorkState.NEW), leftover


def _parse_comments_and_sub_moments(mom: Moment, state: ParseState, indent: int):
    next_indent = indent + state.config.tab_size
    for line in _each_line(state.line_iter):
        line_indent, indent_char_cnt = _count_indent(line.content, state.config.tab_size, next_indent)

        if line_indent >= next_indent:
            _parse_sub_line(mom, line, line.content[indent_char_cnt:], next_indent, state)
        elif _is_blank(line.content):
            if len(mom.comments) > 0:
                # special case: treat empty line between comments as a comment
                mom.comments.append(
                    Comment(
                        content="",
                        doc_pos=DocPosition(line_num=line.line_num, offset=line.offset, length=0),
                    ))
                # Otherwise just ignore the empty line
        else:
            # "Unconsume" the line since it is probably meant for a parent moment up the recursion stack
            state.line_iter.undo()
            break

    # Remove trailing empty comments
    while len(mom.comments) > 0 and mom.comments[-1].content == "":
        mom.comments = mom.comments[:-1]


def _parse_sub_line(mom: Moment, line: Line, line_content: str, indent: int, state: ParseState):
    if line_content.startswith(state.config.left_state_bracket):
        sub_mom = _parse_moment_block(line, line_content, state, indent)
        if sub_mom:
            mom.sub_moments.append(sub_mom)
            return

    # If not a sub moment, assume it's a comment
    _, indent_char_cnt = _count_indent(line.content, state.config.tab_size, indent)
    mom.comments.append(
        Comment(content=line_content,
                doc_pos=DocPosition(
                    line_num=line.line_num,
                    offset=line.offset + indent_char_cnt,
                    length=len(line_content),
                )))


def _parse_date(content, config: ParseConfig) -> datetime | None:
    stripped = content.strip()
    for fmt in config.date_formats:
        try:
            return datetime.strptime(stripped, fmt).astimezone(tz=None)
        except ValueError:
            pass

    return None


def _parse_time(content, config: ParseConfig) -> datetime | None:
    stripped = content.strip()
    try:
        return datetime.strptime(stripped, config.time_format)
    except ValueError:
        pass

    return None


def _count_indent(content, tab_size, max_indent) -> Tuple[int, int]:
    """counts the whitespace indent at the start of the string up to maxIndent.
    Spaces count as 1, Tabs count as tabSize indentation.
    Returns the indentation value and the actual number of whitespace characters."""

    indent = 0
    cnt = 0
    for char in content:
        if char == "\t":
            indent += tab_size
            cnt += 1
        elif char == " ":
            indent += 1
            cnt += 1
        else:
            break

        if indent >= max_indent:
            break

    return indent, cnt


def _each_line(line_iter: LineIterator):
    while True:
        line = next(line_iter, None)
        if line is None:
            break
        yield line


def _is_blank(content: str):
    return not content.strip()


def _count_start_whitespaces(content: str):
    for i, char in enumerate(content):
        if not char.isspace():
            return i
    return len(content)


def _stripped_length(content: str):
    return len(content.strip())

File: core/lib/src/parse.ts
import { addDays, endOfDay, isValid, parse, setDay, startOfDay } from 'date-fns'
import { LineIterator, LineIteratorImpl, StringLineIterator } from './LineIterator'
import {
    Category, createTodos, getNow, Line, Moment, MomentDateTime, ParseConfig, Recurrence, RecurrenceType,
    RecurrenceWithoutDocPos,
    RecurringMoment, SingleMoment, Todos, WorkState
} from './models'
import { epochWeek, each } from './util'

interface ParseState {
    config: ParseConfig
    lineIter: LineIterator
    todos: Todos
}

export function parseMomentsString(content: string, config: ParseConfig): Todos {
    return parseMomentsLines(StringLineIterator(content), config)
}

export function parseMoments(lines: Iterator<string>, config: ParseConfig): Todos {
    return parseMomentsLines(new LineIteratorImpl(lines), config)
}

export function parseMomentsLines(lineIter: LineIterator, config: ParseConfig): Todos {
    const state: ParseState = { config, lineIter, todos: createTodos() }
    for (const line of each(state.lineIter)) {
        parseLine(line, state)
    }
    return state.todos
}

function nextLine(lineIter: LineIterator): Line | null {
    const { value, done } = lineIter.next()
    return done ? null : value
}

function parseLine(line: Line, state: ParseState) {
    if (isBlank(line.content)) {
        return
    }

    if (isCategoryDelimiter(line, state.config)) {
        parseCategoryBlock(state)
    }
    else if (line.content.startsWith(state.config.leftStateBracket)) {
        parseTopMomentBlock(line, state)
    }
}

function parseCategoryBlock(state: ParseState) {
    const catLine = nextLine(state.lineIter)
    if (catLine === null) {
        // Expected a category name after category delimiter
        return
    }

    const lineAfterCat = nextLine(state.lineIter)
    if (!isCategoryDelimiter(lineAfterCat, state.config)) {
        // Expected closing delimiter line
        return
    }

    state.todos.categories.push(parseCategory(catLine, state.config))
}

function parseCategory(line: Line, config: ParseConfig): Category {
    let color, content, priority
    [color, content] = parseCategoryColor(line.content);
    [priority, content] = parsePriority(content, config)
    return {
        name: content.trimStart(),
        color,
        priority,
        docPos: {
            lineNum: line.lineNum,
            offset: line.offset,
            length: line.content.length
        }
    }
}

function parseCategoryColor(content: string): [string, string] {
    const start = content.indexOf('[')

    if (start < 0 || !content.endsWith(']')) {
        return ['', content]
    }

    return [content.slice(start + 1, -1), content.slice(0, start).trim()]
}

function parsePriority(content: string, config: ParseConfig): [number, string] {
    let prio = 0
    for (; prio < content.length; prio++) {
        if (content[content.length - 1 - prio] !== config.priorityMark) {
            break
        }
    }

    return [prio, content.slice(0, content.length - prio).trim()]
}

function isCategoryDelimiter(line: Line | null, config: ParseConfig) {
    return line !== null && line.content.startsWith(config.categoryDelim)
}

function parseTopMomentBlock(line: Line, state: ParseState) {
    const mom = parseMomentBlock(line, line.content.trim(), state)
    if (!mom) {
        return
    }
    state.todos.moments.push(mom)
}

function parseMomentBlock(line: Line, lineContent: string, state: ParseState, indent = 0): Moment | null {
    const mom = parseMomentLine(line, lineContent, state.config)
    if (!mom) {
        return null
    }

    mom.category = state.todos.categories ? state.todos.categories[state.todos.categories.length - 1] : null
    parseCommentsAndSubMoments(mom, state, indent)
    return mom
}

function parseMomentLine(line: Line, lineContent: string, config: ParseConfig): Moment | null {
    let mom: Partial<Moment> | null;
    [mom, lineContent] = parseRecurringMoment(line, lineContent, config)
    if (!mom) {
        [mom, lineContent] = parseSingleMoment(line, lineContent, config)
    }

    let state
    [state, lineContent] = parseStateMark(lineContent, config)
    if (state === null) {
        return null
    }
    mom.workState = state;

    [mom.priority, lineContent] = parsePriority(lineContent, config)

    mom.name = lineContent
    mom.comments = []
    mom.subMoments = []

    return mom as Moment
}

function parseRecurringMoment(line: Line, lineContent: string, config: ParseConfig): [Partial<RecurringMoment> | null, string] {
    if (!lineContent.endsWith(config.rightDateBracket)) {
        return [null, lineContent]
    }

    const [recur, timeOfDay, newLineContent] = parseRecurrence(line, lineContent, config)
    if (!recur) {
        return [null, lineContent]
    }

    const mom: Partial<RecurringMoment> = {
        recurrence: recur,
        timeOfDay,
        docPos: {
            lineNum: line.lineNum,
            offset: line.offset,
            length: line.content.length
        }
    }

    return [mom, newLineContent]
}

function parseRecurrence(line: Line, lineContent: string, config: ParseConfig): [Recurrence | null, MomentDateTime | null, string] {
    const lbracketPos = lineContent.lastIndexOf(config.leftDateBracket)
    if (lbracketPos < 0) {
        return [null, null, lineContent]
    }

    const untrimmedPos = line.content.lastIndexOf(config.leftDateBracket) + 1
    let recurStr = lineContent.slice(lbracketPos + 1, -1)
    let timeOfDay
    [timeOfDay, recurStr] = parseTimeSuffix(recurStr, config)
    if (timeOfDay) {
        timeOfDay.docPos.offset += line.offset + untrimmedPos
    }

    let recurWithoutDocPos = null
    for (const parseFunc of [tryParseDaily, tryParseWeekly, tryParseNWeekly, tryParseMonthly, tryParseYearly]) {
        recurWithoutDocPos = parseFunc(recurStr, config)
        if (recurWithoutDocPos) {
            break
        }
    }

    if (!recurWithoutDocPos) {
        return [null, null, lineContent]
    }

    const recur = {
        ...recurWithoutDocPos,
        refDate: {
            ...recurWithoutDocPos.refDate,
            docPos: {
                lineNum: line.lineNum,
                offset: line.offset + untrimmedPos,
                length: recurStr.length
            }
        }
    }

    return [recur, timeOfDay, lineContent.slice(0, lbracketPos).trim()]
}

function tryParseDaily(content: string, config: ParseConfig): RecurrenceWithoutDocPos | null {
    if (!content.match(config.dailyPattern)) {
        return null
    }

    return {
        recurrenceType: RecurrenceType.DAILY,
        refDate: { dt: getNow(config) }
    }
}

function tryParseWeekly(content: string, config: ParseConfig): RecurrenceWithoutDocPos | null {
    const match = content.match(config.weeklyPattern)
    if (!match) {
        return null
    }

    const weekday = parseWeekday(match[1], config)
    if (weekday < 0) {
        return null
    }

    const refDate = setDay(getNow(config), weekday)
    return {
        recurrenceType: RecurrenceType.WEEKLY,
        refDate: { dt: startOfDay(refDate) }
    }
}

function parseWeekday(content: string, config: ParseConfig): number {
    return config.weekDays.indexOf(content.toLowerCase())
}

function tryParseNWeekly(content: string, config: ParseConfig): RecurrenceWithoutDocPos | null {
    const match = content.match(config.nWeeklyPattern)
    if (!match) {
        return null
    }

    const [nth, nthRecurType] = parseNth(match[1], config)
    if (!nthRecurType) {
        return null
    }

    const weekday = parseWeekday(match[2], config)
    if (weekday < 0) {
        return null
    }

    let refDate = setDay(getNow(config), weekday)
    const weekOffset = epochWeek(refDate) % nth
    refDate = addDays(refDate, -7 * weekOffset)
    return {
        recurrenceType: nthRecurType,
        refDate: { dt: startOfDay(refDate) }
    }
}

function parseNth(content: string, config: ParseConfig): [number, RecurrenceType | null] {
    switch (config.nths.indexOf(content.toLowerCase())) {
        case 0:
            return [2, RecurrenceType.BIWEEKLY]
        case 1:
            return [3, RecurrenceType.TRIWEEKLY]
        case 2:
            return [4, RecurrenceType.QUADRIWEEKLY]
        default:
            return [-1, null]
    }
}

function tryParseMonthly(content: string, config: ParseConfig): RecurrenceWithoutDocPos | null {
    const match = content.match(config.monthlyPattern)
    if (!match) {
        return null
    }

    const day = parseInt(match[1])
    if (isNaN(day)) {
        return null
    }

    const now = getNow(config)
    const refDate = new Date(now.getFullYear(), now.getMonth(), day)

    return {
        recurrenceType: RecurrenceType.MONTHLY,
        refDate: { dt: refDate }
    }
}

function tryParseYearly(content: string, config: ParseConfig): RecurrenceWithoutDocPos | null {
    const match = content.match(config.yearlyPattern)
    if (!match) {
        return null
    }

    const day = parseInt(match[1])
    const month = parseInt(match[2])
    if (isNaN(day) || isNaN(month)) {
        return null
    }

    const now = getNow(config)
    const refDate = new Date(now.getFullYear(), month - 1, day)
    return {
        recurrenceType: RecurrenceType.YEARLY,
        refDate: { dt: refDate }
    }
}

function parseSingleMoment(line: Line, lineContent: string, config: ParseConfig): [Partial<SingleMoment>, string] {
    let start = null
    let end = null
    let timeOfDay = null
    if (lineContent.endsWith(config.rightDateBracket)) {
        [start, end, timeOfDay, lineContent] = parseDateSuffix(line, lineContent, config)
    }

    if (end) {
        end.dt = endOfDay(end.dt)
    }

    return [
        {
            start,
            end,
            timeOfDay,
            docPos: {
                lineNum: line.lineNum,
                offset: line.offset,
                length: line.content.length
            }
        },
        lineContent
    ]
}

function parseDateSuffix(line: Line, lineContent: string, config: ParseConfig): [MomentDateTime | null, MomentDateTime | null, MomentDateTime | null, string] {
    const leftPos = lineContent.lastIndexOf(config.leftDateBracket)
    if (leftPos < 0) {
        return [null, null, null, lineContent]
    }

    const untrimmedPos = line.content.lastIndexOf(config.leftDateBracket) + 1
    let suffixStr = lineContent.slice(leftPos + 1, -1)
    let timeOfDay
    [timeOfDay, suffixStr] = parseTimeSuffix(suffixStr, config)
    finalizeDocPos(timeOfDay, line, untrimmedPos)

    suffixStr = suffixStr.trim()
    let start: MomentDateTime | null = null
    let end: MomentDateTime | null = null
    // Try all dashes as separator in case the date format uses dashes (e.g. 2015-12-24 - 2016-02-03)
    let dashOffset = 0
    let dashPos = suffixStr.indexOf('-')
    while (dashPos >= 0 && start === null) {
        [start, end] = parseDateRange(suffixStr, dashOffset + dashPos, config)
        if (start == null) {
            dashOffset += dashPos + 1
            dashPos = suffixStr.indexOf('-', dashOffset)
        }
    }

    if (!start) {
        start = parseDateSingle(suffixStr, config)
        if (start) {
            end = { ...start }
            end.docPos = { ...start.docPos }
        }
    }

    if (end) {
        end.dt = endOfDay(end.dt)
    }

    if (start || end) {
        finalizeDocPos(start, line, untrimmedPos)
        finalizeDocPos(end, line, untrimmedPos)
        return [start, end, timeOfDay, lineContent.slice(0, leftPos).trim()]
    }

    return [null, null, null, lineContent]
}

function parseDateRange(lineContent: string, dashPos: number, config: ParseConfig): [MomentDateTime | null, MomentDateTime | null] {
    let start: MomentDateTime | null = null
    let end: MomentDateTime | null = null
    const startStr = lineContent.slice(0, dashPos)
    const endStr = lineContent.slice(dashPos + 1)

    if (startStr) {
        const date = parseDate(startStr, config)
        if (!date) {
            return [null, null]
        }
        start = {
            dt: startOfDay(date),
            docPos: {
                lineNum: -1,
                offset: countStartWhitespaces(startStr),
                length: strippedLength(startStr)
            }
        }
    }

    if (endStr) {
        const date = parseDate(endStr, config)
        if (!date) {
            return [null, null]
        }
        end = {
            dt: endOfDay(date),
            docPos: {
                lineNum: -1,
                offset: startStr.length + 1 + countStartWhitespaces(endStr),
                length: strippedLength(endStr)
            }
        }
    }

    return [start, end]
}

function parseDateSingle(lineContent: string, config: ParseConfig): MomentDateTime | null {
    const date = parseDate(lineContent, config)

    if (!date) {
        return null
    }

    return {
        dt: startOfDay(date),
        docPos: {
            lineNum: -1,
            offset: countStartWhitespaces(lineContent),
            length: strippedLength(lineContent)
        }
    }
}

function parseTimeSuffix(lineContent: string, config: ParseConfig): [MomentDateTime | null, string] {
    const trimmed = lineContent.trim()
    const spacePos = trimmed.lastIndexOf(' ')
    if (spacePos < 0 || spacePos === trimmed.length - 1) {
        return [null, lineContent]
    }

    const timeStr = trimmed.slice(spacePos + 1)
    const timeOfDay = parseTime(timeStr, config)
    if (!timeOfDay) {
        return [null, lineContent]
    }

    const momTime = {
        dt: timeOfDay,
        docPos: {
            lineNum: -1,
            offset: spacePos + 1,
            length: timeStr.length
        }
    }

    return [momTime, lineContent.slice(0, spacePos)]
}

function finalizeDocPos(date: MomentDateTime | null, line: Line, offsetDelta: number) {
    if (date) {
        date.docPos.lineNum = line.lineNum
        date.docPos.offset += line.offset + offsetDelta
    }
}

function parseStateMark(lineContent: string, config: ParseConfig): [WorkState | null, string] {
    const match = config.stateMarkPattern.exec(lineContent)
    if (!match) {
        return [null, lineContent]
    }
    const leftover = lineContent.slice(match.index + match[0].length + 1).trim()
    return [config.stateMarks[match[1]] ?? WorkState.NEW, leftover]
}

function parseCommentsAndSubMoments(mom: Moment, state: ParseState, indent: number) {
    const nextIndent = indent + state.config.tabSize

    for (const line of each(state.lineIter)) {
        const [lineIndent, indentCharCnt] = countIndent(line.content, state.config.tabSize, nextIndent)

        if (lineIndent >= nextIndent) {
            parseSubLine(mom, line, line.content.slice(indentCharCnt), nextIndent, state)
        }
        else if (isBlank(line.content)) {
            if (mom.comments.length > 0) {
                // special case: treat empty line between comments as a comment
                mom.comments.push(
                    {
                        content: '',
                        docPos: { lineNum: line.lineNum, offset: line.offset, length: 0 }
                    })
            }
        }
        // Otherwise just ignore the empty line
        else {
            // "Unconsume" the line since it is probably meant for a parent moment up the recursion stack
            state.lineIter.undo()
            break
        }
    }

    // Remove trailing empty comments
    while (mom.comments.length > 0 && mom.comments[mom.comments.length - 1].content === '') {
        mom.comments = mom.comments.slice(0, -1)
    }
}

function parseSubLine(mom: Moment, line: Line, lineContent: string, indent: number, state: ParseState) {
    if (lineContent.startsWith(state.config.leftStateBracket)) {
        const subMom = parseMomentBlock(line, lineContent, state, indent)
        if (subMom) {
            mom.subMoments.push(subMom)
            return
        }
    }

    // If not a sub moment, assume it's a comment
    const [, indentCharCnt] = countIndent(line.content, state.config.tabSize, indent)
    mom.comments.push(
        {
            content: lineContent,
            docPos: {
                lineNum: line.lineNum,
                offset: line.offset + indentCharCnt,
                length: lineContent.length
            }
        })
}

function parseDate(content: string, config: ParseConfig): Date | null {
    const stripped = content.trim()
    for (const fmt of config.dateFormats) {
        const dt = parse(stripped, fmt, new Date())
        if (isValid(dt)) {
            return dt
        }
    }

    return null
}

function parseTime(content: string, config: ParseConfig): Date | null {
    const stripped = content.trim()
    const dt = parse(stripped, config.timeFormat, new Date())
    return isValid(dt) ? dt : null
}

/**
 * counts the whitespace indent at the start of the string up to maxIndent.
 * Spaces count as 1, Tabs count as tabSize indentation.
 * Returns the indentation value and the actual number of whitespace characters.
*/
function countIndent(content: string, tabSize: number, maxIndent: number): [number, number] {
    let indent = 0
    let cnt = 0
    for (const char of content) {
        if (char === '\t') {
            indent += tabSize
            cnt += 1
        }
        else if (char === ' ') {
            indent += 1
            cnt += 1
        }
        else {
            break
        }

        if (indent >= maxIndent) {
            break
        }
    }

    return [indent, cnt]
}

function isBlank(content: string) {
    return !content.trim()
}

function countStartWhitespaces(content: string) {
    for (let i = 0; i < content.length; i++) {
        if (!/\s/.test(content[i])) {
            return i
        }
    }

    return content.length
}

function strippedLength(content: string) {
    return content.trim().length
}
